### PR TITLE
Nft sign + execute

### PR DESCRIPTION
**Pull request by mikiquantum**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Issue Link: <Github link>

### CODE REVIEW

File: identity/did.go [hunk 1]
```diff
@@ -198,10 +199,10 @@ type ServiceDID interface {
 	GetKey(did DID, key [32]byte) (*KeyResponse, error)
 
 	// RawExecute calls the execute method on the identity contract
-	RawExecute(ctx context.Context, to common.Address, data []byte) error
+	RawExecute(ctx context.Context, to common.Address, data []byte) (utxID uuid.UUID, done chan bool, err error)
```
Comment (mikiquantum) [inline identity/did.go]: Cyclic dependency madness with transactions package. Only way could think about fixing it fast and dirty. Explanation in the implementation.

File: identity/ideth/service.go [hunk 2]
```diff
@@ -242,45 +244,37 @@ func (i service) IsSignedWithPurpose(did id.DID, message [32]byte, signature []b
...
-
-	isDone := <-done
-	// non async task
-	if !isDone {
-		return errors.New("raw execute TX failed: txID:%s", txID.String())
-
-	}
-	return nil
-
+	txID, done, err = i.txManager.ExecuteWithinTX(context.Background(), DID, txID, "Check TX for execute", i.ethereumTX(opts, contract.Execute, to, value, data))
```
Comment (mikiquantum) [inline identity/ideth/service.go]: Convert it to non-blocking.

File: nft/ethereum_payment_obligation.go [hunk 3]
```diff
@@ -217,18 +199,15 @@ func (s *ethereumPaymentObligation) minter(ctx context.Context, tokenID TokenID,
 		log.Debugf("Salts: %s", byte32SlicetoString(requestData.Salts))
 		log.Debugf("Proofs: %s", byteByte32SlicetoString(requestData.Proofs))
 
-		res, err := ethereum.QueueEthTXStatusTask(accountID, txID, ethTX.Hash(), s.queue)
-		if err != nil {
-			errOut <- err
+		isDone = <-done
```
Comment (mikiquantum) [inline nft/ethereum_payment_obligation.go]: moved the blocking up to the invoker.

File: nft/ethereum_payment_obligation.go [hunk 4]
```diff
@@ -309,8 +288,8 @@ func convertToProofData(proofspb []*proofspb.Proof) (*proofData, error) {
 	var proofs = make([][][32]byte, len(proofspb))
 
 	// TODO remove later
-	//proof, _ := documents.ConvertDocProofToClientFormat(&documents.DocumentProof{FieldProofs: proofspb})
-	//log.Info(json.MarshalIndent(proof, "", "  "))
+	proof, _ := documents.ConvertDocProofToClientFormat(&documents.DocumentProof{FieldProofs: proofspb})
```
Comment (mikiquantum) [inline nft/ethereum_payment_obligation.go]: will comment this, as soon as the new contracts are integrated successfully.

File: documents/coredocument.go [hunk 5]
```diff
@@ -276,9 +284,10 @@ func (cd *CoreDocument) CreateProofs(docType string, dataTree *proofs.DocumentTr
 		return nil, err
 	}
 
-	treeProofs[dataPrefix] = newTreeProof(dataTree, append([][]byte{cdRoot}, srpHashes...))
-	treeProofs[SignaturesTreePrefix] = newTreeProof(signatureTree, srpHashes)
-	treeProofs[CDTreePrefix] = newTreeProof(cdTree, append([][]byte{dataRoot}, srpHashes...))
+	treeProofs[DRTreePrefix] = newTreeProof(drTree, nil)
```
Comment (mikiquantum) [inline documents/coredocument.go]: Fixed tree hashes

File: documents/coredocument.go [hunk 6]
```diff
@@ -416,16 +433,14 @@ func (cd *CoreDocument) signingRootTree(docType string) (tree *proofs.DocumentTr
 
 	// create the signing tree with data root and coredoc root as siblings
 	tree = NewDefaultTreeWithPrefix(ConvertToProofSalts(cd.Document.CoredocumentSalts), SigningTreePrefix, compactProperties(SigningTreePrefix))
-	prefixProp := NewLeafProperty(SigningTreePrefix, compactProperties(SigningTreePrefix))
-
 	err = tree.AddLeaves([]proofs.LeafNode{
 		{
-			Property: prefixProp.FieldProp(DataRootField, binary.LittleEndian.Uint32(compactProperties(DataRootField))),
+			Property: NewLeafProperty(fmt.Sprintf("%s.%s", SigningTreePrefix, DataRootField), append(compactProperties(SigningTreePrefix), compactProperties(DataRootField)...)),
```
Comment (mikiquantum) [inline documents/coredocument.go]: avoid little endian encoding, and letting the precise proof library take care of that.

File: identity/did.go (hunk 1 above)
Comment (vimukthi-git) [reply identity/did.go]: There is a fast and clean way to do it as done by other packages already, just define an interface that TxID automatically implements in this file and return a pointer to TxID type in the ideth package :) 
```
type IDTX interface {
	String() string
	Bytes() []byte
}
```
Thanks!

File: identity/ideth/service.go [hunk 7]
```diff
@@ -230,68 +234,57 @@ func (i service) GetKey(did id.DID, key [32]byte) (*id.KeyResponse, error) {
...
-
 // RawExecute calls the execute method on the identity contract
-func (i service) RawExecute(ctx context.Context, to common.Address, data []byte) error {
+// TODO once we clean up transaction to not use higher level deps we can change back the return to be transactions.txID
+func (i service) RawExecute(ctx context.Context, to common.Address, data []byte) (txID id.IDTX, done chan bool, err error) {
+	utxID := contextutil.TX(ctx)
 	DID, err := NewDIDFromContext(ctx)
 	if err != nil {
-		return err
+		return uuid.Nil, nil, err
```
Comment (vimukthi-git) [inline identity/ideth/service.go]: `TxID.NilTxID() `

File: identity/ideth/service.go [hunk 8]
```diff
@@ -230,68 +234,57 @@ func (i service) GetKey(did id.DID, key [32]byte) (*id.KeyResponse, error) {
...
 	}
 	contract, opts, err := i.prepareTransaction(ctx, DID)
 	if err != nil {
-		return err
+		return uuid.Nil, nil, err
```
Comment (vimukthi-git) [inline identity/ideth/service.go]: same here

File: identity/ideth/service.go [hunk 9]
```diff
@@ -7,6 +7,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/satori/go.uuid"
```
Comment (vimukthi-git) [inline identity/ideth/service.go]: remove

**Pull request merged**